src/main/java/appeng/me/storage/CellInventoryBase.java

```diff
@@ -0,0 +1,374 @@
+	protected final ISaveProvider container;
+	private int maxItemTypes = 63;
+	private short storedItems = 0;
+	private int storedItemCount = 0;
+	protected IItemList<T> cellItems;
+	protected ItemStack i;
+	protected IStorageCell cellType;
+	private final int itemsPerByte;
+
+	protected CellInventoryBase( final NBTTagCompound data, final ISaveProvider container, final int itemsPerByte )
+	{
+		this.tagCompound = data;
+		this.container = container;
+		this.itemsPerByte = itemsPerByte;
+	}
+
+	protected CellInventoryBase( final ItemStack o, final ISaveProvider container, final int itemsPerByte ) throws AppEngException
+	{
+		this.itemsPerByte = itemsPerByte;
+		if( itemSlots == null )
+		{
+			itemSlots = new String[this.maxItemTypes];
+			itemSlotCount = new String[this.maxItemTypes];
+
+			for( int x = 0; x < this.maxItemTypes; x++ )
+			{
+				itemSlots[x] = ITEM_SLOT + x;
+				itemSlotCount[x] = ITEM_SLOT_COUNT + x;
+			}
+		}
+
+		if( o == null )
+		{
+			throw new AppEngException( "ItemStack was used as a cell, but was not a cell!" );
+		}
+
+		this.cellType = null;
+		this.i = o;
+
+		final Item type = this.i.getItem();
+		if( type instanceof IStorageCell )
+		{
+			this.cellType = (IStorageCell) this.i.getItem();
+			this.maxItemTypes = this.cellType.getTotalTypes( this.i );
+		}
+
+		if( this.cellType == null )
+		{
+			throw new AppEngException( "ItemStack was used as a cell, but was not a cell!" );
+		}
+
+		if( !this.cellType.isStorageCell( this.i ) )
+		{
+			throw new AppEngException( "ItemStack was used as a cell, but was not a cell!" );
+		}
+
+		if( this.maxItemTypes > 63 )
+		{
+			this.maxItemTypes = 63;
+		}
+		if( this.maxItemTypes < 1 )
+		{
+			this.maxItemTypes = 1;
+		}
+
+		this.container = container;
+		this.tagCompound = Platform.openNbtData( o );
+		this.storedItems = this.tagCompound.getShort( ITEM_TYPE_TAG );
+		this.storedItemCount = this.tagCompound.getInteger( ITEM_COUNT_TAG );
+		this.cellItems = null;
+	}
+
+	protected boolean isEmpty( final IMEInventory meInventory )
+	{
+		return meInventory.getAvailableItems( getChannel().createList() ).isEmpty();
+	}
+
+	protected IItemList<T> getCellItems()
+	{
+		if( this.cellItems == null )
+		{
+			this.cellItems = getChannel().createList();
+			this.loadCellItems();
+		}
+
+		return this.cellItems;
+	}
+
+	protected void updateItemCount( final long delta )
+	{
+		this.storedItemCount += delta;
+		this.tagCompound.setInteger( ITEM_COUNT_TAG, this.storedItemCount );
+	}
+
+	protected void saveChanges()
+	{
+		// cellItems.clean();
+		int itemCount = 0;
+
+		// add new pretty stuff...
+		int x = 0;
+		for( final T v : this.cellItems )
+		{
+			itemCount += v.getStackSize();
+
+			final NBTTagCompound g = new NBTTagCompound();
+			v.writeToNBT( g );
+			this.tagCompound.setTag( itemSlots[x], g );
+
+			this.tagCompound.setInteger( itemSlotCount[x], (int) v.getStackSize() );
+
+			x++;
+		}
+
+		// NBTBase tagType = tagCompound.getTag( ITEM_TYPE_TAG );
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') This block of commented-out lines of code should be removed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)